src/ds/thread_messaging.h

```diff
@@ -0,0 +1,210 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the Apache 2.0 License.
+#pragma once
+
+//#define USE_MPSCQ
+
+#include "ds/logger.h"
+#ifdef USE_MPSCQ
+#  include "ds/mpscq.h"
+#endif
+#include "ds/ts.h"
+
+#include <atomic>
+#include <cstddef>
+
+namespace enclave
+{
+  struct ThreadMsg
+  {
+    void (*cb)(std::unique_ptr<ThreadMsg>);
+    std::atomic<ThreadMsg*> next = nullptr;
+    uint64_t padding[14];
+  };
+
+  template <typename Payload>
+  struct Tmsg
+  {
+    Tmsg(void (*_cb)(std::unique_ptr<Tmsg<Payload>>)) :
+      cb(reinterpret_cast<void (*)(std::unique_ptr<ThreadMsg>)>(_cb)),
+      next(nullptr)
+    {
+      check_invariants();
+    }
+
+    void (*cb)(std::unique_ptr<ThreadMsg>);
+    std::atomic<ThreadMsg*> next;
+    union
+    {
+      Payload data;
+      uint64_t padding[14];
+    };
+
+    static void check_invariants()
+    {
+      static_assert(
+        sizeof(ThreadMsg) == sizeof(Tmsg<Payload>), "message is too large");
+      static_assert(
+        sizeof(Payload) <= sizeof(ThreadMsg::padding),
+        "message payload is too large");
+
+      static_assert(
+        offsetof(Tmsg, cb) == offsetof(ThreadMsg, cb),
+        "Expected cb at start of struct");
+      static_assert(
+        offsetof(Tmsg, next) == offsetof(ThreadMsg, next),
+        "Expected next after cb in struct");
+      static_assert(
+        offsetof(Tmsg, data) == offsetof(ThreadMsg, padding),
+        "Expected payload after next in struct");
+    }
+  };
+
+  static void init_cb(std::unique_ptr<ThreadMsg> stuff)
```

Comment:
As discussed, this can go.

Reply:
Sorry i missed this but it is used to in the the MPSCQ